### ISSUE TITLE
`editor.range =` vs `editor.selectRange` vs `postEditor.setRange()`?

### ISSUE THREAD

**Issue by sdhull** [bustle/mobiledoc-kit#639]
Been looking at these APIs and not sure which I should use in which circumstances. A little guidance would be appreciated.
Thanks!

**Comment by sdhull** [comment]
Tracing these methods, it appears that `editor.selectRange` calls `cursor.selectRange` & `editor.range =` (eg, the editor's `set range` method). So I probably shouldn't call `editor.range =` directly.

However, as far as I can tell, `editor.selectRange` never calls `postEditor.setRange`?  So do I need to do that additionally?

**Comment by kevinansfield** [comment]
@sdhull my understanding is that the difference comes down to whether you are inside an `editor.run` callback or not.

`editor.selectRange` is used outside of an `editor.run` callback as it changes the range and fires the `cursorDidChange` etc hooks immediately. Useful when you have application code that wants to change the cursor position without performing any other edits.

`postEditor.setRange` is used inside of an `editor.run` call (this is when you get access to the `postEditor` instance) because it instead only updates the internal edit state, scheduling the cursor change hooks to fire once the programmatic edit is finished. That way you can perform multiple cursor changes without any side effects caused by hooks firing on intermediate positions.

I've never personally set the range directly with `editor.range =` within application code.

**Comment by sdhull** [comment]
@kevinansfield thanks for the reply! That kinda makes sense I guess. The thing I'm still struggling with is why I can't do `editor.toggleMarkup` after setting the range to a collapsed range (eg, a cursor).

Scenario is:
* user is typing, wants to add a custom markup
* clicks in the toolbar to select custom markup by typeahead, however that typeahead steals focus
  * so we cache the current range when they click the toolbar
  * when they choose the markup, we restore the range and do `editor.toggleMarkup`
* result = if the range was not collapsed (eg, an actual selection was present), `toggleMarkup` works great, if the range _was_ collapsed (eg, just a cursor), `toggleMarkup` fails silently, however the cursor is restored and user can type in the doc again

So I've been guessing maybe I'm not restoring the range properly?

**Comment by kevinansfield** [comment]
Hmm, that sounds like a similar issue that I've had to [work around before](https://github.com/TryGhost/Ghost-Admin/blob/master/lib/koenig-editor/addon/options/text-expansions.js#L272-L275). The problem there was that after the cursor movement/programmatic edit, mobiledoc-kit would use the DOM state to update it's `activeMarkupState` which reversed the toggle that had just been made because the cursor was still touching a markup. My fix was to schedule the toggle to happen after mobiledoc-kit's internal housekeeping had run. Does that sound similar to what you're dealing with?

**Comment by sdhull** [comment]
After stepping through the method (with just a cursor), I found it does make it through the code path to do `editState._addActiveMarkup()`

**Comment by sdhull** [comment]
Yes I'll try that and report back... luckily we're using Ember too 😀

**Comment by sdhull** [comment]
OK scheduling in the actions queue did not work (eg, [`once`](https://emberjs.com/api/ember/2.18/functions/@ember%2Frunloop/once)), however using [`next`](https://emberjs.com/api/ember/2.18/functions/@ember%2Frunloop/next) does work.

It feels a bit precarious... it's essentially the same as using `sleep` -- I would much rather have a callback I could use that would guarantee that mobiledoc-kit had finished whatever internal housekeeping is necessary to make that toggle work before actually calling it. Any ideas if there is such a callback I could attach to?

**Comment by sdhull** [comment]
ugh after testing some more, `next()` doesn't work reliably. It works about 50% of the time.

I tried `editor.addCallbackOnce()` (undocumented) but none of the hooks I tried worked properly.
Short of a hook I could use in editor, if there was some way to introspect the state to determine whether or not it was ready to toggle the markup, that would be acceptable as well. Then I could basically keep doing `next()` until it was ready
